src/org/sosy_lab/java_smt/api/UserPropagator.java

```diff
@@ -0,0 +1,114 @@
+public interface UserPropagator {
+
+  /**
+   * This callback is invoked whenever the solver creates a new decision level.
+   * A user propagator should maintain backtracking points on each push to enable later
+   * backtracking.
+   */
+  void onPush();
+
+  /**
+   * This callback is invoked when the solver backtracks. The solver can backtrack multiple
+   * levels simultaneously.
+   * @param numPoppedLevels The number of levels to backtrack (~ number of pushes to backtrack).
+   */
+  void onPop(int numPoppedLevels);
+
+  /**
+   * This callback is invoked when the solver finds a complete satisfying assignment.
+   * The user can check the found model for consistency and potentially raise conflicts via
+   * {@link PropagatorBackend#propagateConflict(BooleanFormula[])}.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnFinalCheck()}.
+   */
+  void onFinalCheck();
+
+  /**
+   * This callback is invoked if the solver gets to know the value of a registered expression
+   * ({@link #registerExpression}).
+   * Within the callback, the user can raise conflicts via
+   * {@link PropagatorBackend#propagateConflict}, propagate consequences via
+   * {@link PropagatorBackend#propagateConsequence}, or influence the solvers decision heuristics
+   * via {@link PropagatorBackend#propagateNextDecision}.
+   * <p>
+   * The reported value is only known on the current and later push levels,
+   * but will get invalidated when backtracking.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnKnownValue()}.
+   *
+   * @param expr  The expressions whose value is known.
+   * @param value The value of the expression.
+   */
+  void onKnownValue(BooleanFormula expr, boolean value);
+
+  /**
+   * This callback is invoked if the solver decides to branch on a registered expression.
+   * ({@link #registerExpression}).
+   * Within the callback, the user can change the decision by calling
+   * {@link PropagatorBackend#propagateNextDecision}.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnDecision()}.
+   * @param expr The expressions whose value gets decided (usually a literal).
+   * @param value The decision value.
+   */
+  void onDecision(BooleanFormula expr, boolean value);
+
+  /**
+   * Connects this user propagator with a {@link PropagatorBackend}. The backend is used
+   * to register expressions, raise conflicts, propagate consequences, etc.
+   * @param backend The propagator backend.
+   */
+  void injectBackend(PropagatorBackend backend);
+
+  /**
+   * This method is similar to a constructor but is guaranteed to get invoked only after
+   * {@link #injectBackend} was successfully called.
+   * The user can enable notifications by accessing the injected {@link PropagatorBackend}.
+   */
+  void initialize();
```

Comment:
Would it make sense to inline th method "initialize" into "injectBackend" directly to simplify the user-directed API and name both together as "initializeWithBackend(backend)"?

Reply:
Hmm, I separated those two to make it easier for the user to avoid any mistakes, since there is basically only a single reasonable implementation for `injectBackend` and that one can be inherited from `AbstractUserPropagator`.
I can also merge them, if you think that is better.

Reply:
I combined both methods now.